feat: add C ndarray interface and refactor implementation for `stats/base/dnanvarianceyc`

0PrashantYadav0 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: na
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: passed
  - task: lint_c_examples status: passed
  - task: lint_c_benchmarks status: passed
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: na
  - task: lint_license_headers status: passed ---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes. report:
  - task: run_javascript_examples status: na
  - task: run_c_examples status: na
  - task: run_cpp_examples status: na
  - task: run_javascript_readme_examples status: na
  - task: run_c_benchmarks status: na
  - task: run_cpp_benchmarks status: na
  - task: run_fortran_benchmarks status: na
  - task: run_javascript_benchmarks status: na
  - task: run_julia_benchmarks status: na
  - task: run_python_benchmarks status: na
  - task: run_r_benchmarks status: na
  - task: run_javascript_tests status: na ---

Resolves None.

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C ndarray interface for `stats/base/dnanvarianceyc`
-   refactors JavaScript implementation
-   updates tests, docs, examples and benchmarks

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None
## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

0PrashantYadav0 committed:
feat: add C ndarray interface and refactor implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dnanvarianceyc/README.md:
```diff
@@ -137,44 +135,39 @@ Note that indexing is relative to the first index. To introduce an offset, use [
 var Float64Array = require( '@stdlib/array/float64' );
 
-var x = new Float64Array( [ 1.0, -2.0, NaN, 2.0 ] );
+var x = new Float64Array( [ 1.0, -2.0, 2.0 ] );
```
```suggestion
var x = new Float64Array( [ 1.0, -2.0, NaN, 2.0 ] );
```
we should show how the `NaN` element is being treated by the example, no need to remove it

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dnanvarianceyc/examples/index.js:
```diff
@@ -18,23 +18,19 @@
 }
-console.log( x );
+
+var x = filledarrayBy( 10, 'float64', rand );
```
```suggestion
var x = filledarrayBy( 10, 'float64', rand );
console.log( x );
```

let's not remove the console.log

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dnanvarianceyc/lib/dnanvarianceyc.js:
```diff
@@ -46,62 +52,8 @@
 * var v = dnanvarianceyc( N, 1, x, 1 );
```
```suggestion
* var v = dnanvarianceyc( x.length, 1, x, 1 );
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dnanvarianceyc/src/addon.c:
```diff
@@ -1,7 +1,7 @@
 * @license Apache-2.0
 *
-* Copyright (c) 2025 The Stdlib Authors.
+* Copyright (c) 2024 The Stdlib Authors.
```
```suggestion
* Copyright (c) 2025 The Stdlib Authors.
```
no need for changes here

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dnanvarianceyc/src/addon.c:
```diff
@@ -36,10 +37,28 @@ static napi_value addon( napi_env env, napi_callback_info info ) {
-	STDLIB_NAPI_ARGV_STRIDED_FLOAT64ARRAY( env, X, N, stride, argv, 2 );
-	STDLIB_NAPI_CREATE_DOUBLE( env, stdlib_strided_dnanvarianceyc( N, correction, X, stride ), v );
+	STDLIB_NAPI_ARGV_INT64( env, strideX, argv, 3 );
+	STDLIB_NAPI_ARGV_STRIDED_FLOAT64ARRAY( env, X, N, strideX, argv, 2 )
```
```suggestion
	STDLIB_NAPI_ARGV_STRIDED_FLOAT64ARRAY( env, X, N, strideX, argv, 2 );
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dnanvarianceyc/src/main.c:
```diff
@@ -34,45 +35,57 @@
 */
-double stdlib_strided_dnanvarianceyc( const int64_t N, const double correction, const double *X, const int64_t stride ) {
+double API_SUFFIX(stdlib_strided_dnanvarianceyc)( const CBLAS_INT N, const double correction, const double *X, const CBLAS_INT strideX ) {
+	const CBLAS_INT ox = stdlib_strided_stride2offset( N, strideX );
+	return API_SUFFIX(stdlib_strided_dnanvarianceyc_ndarray)( N, correction, X, strideX, ox );
+}
+
+/**
+* Computes the variance of a double-precision floating-point strided array ignoring `NaN` values and using a one-pass algorithm proposed by Youngs and Cramer and alternative indexing semantics.
+*
+* @param N            number of indexed elements
+* @param correction   degrees of freedom adjustment
+* @param X            input array
+* @param strideX      stride length
+* @param offsetX      starting index for X
+* @return             variance
+*/
+double API_SUFFIX(stdlib_strided_dnanvarianceyc_ndarray)( const CBLAS_INT N, const double correction, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
 	double sum;
-	int64_t ix;
+	CBLAS_INT ix;
+	CBLAS_INT i;
```
please order these by length

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dnanvarianceyc/src/main.c:
```diff
@@ -34,45 +35,57 @@
 * @param correction  degrees of freedom adjustment
 * @param X           input array
 * @param stride      stride length
```
```suggestion
* @param strideX      stride length
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dnanvarianceyc/src/main.c:
```diff
@@ -34,45 +35,57 @@
 * @param X           input array
 * @param stride      stride length
-* @return            output value
+* @return            variance
```
```suggestion
* @return            output value
```
no need for changes here IMO

0PrashantYadav0 committed:
chore: updated according to code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 commented:
/stdlib help

aayush0325 commented:
can you also make changes in examples similar to what is done in #4994, this would show better handling of `NaN` elements

0PrashantYadav0 committed:
chore: added NaN value in array

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aayush0325 reviewed (dismissed):
thanks!

0PrashantYadav0 commented:
@kgryte Please review and merge this one too. Thank you.

kgryte commented on lib/node_modules/@stdlib/stats/base/dnanvarianceyc/README.md:
```diff
@@ -163,18 +160,18 @@ var v = dnanvarianceyc.ndarray( x.length, 1, x, 1, 0 );
 
-var x = new Float64Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
-var N = floor( x.length / 2 );
+var x = new Float64Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0, NaN, NaN] );
```
```suggestion
var x = new Float64Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0, NaN, NaN ] );
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dnanvarianceyc/README.md:
```diff
@@ -222,6 +220,125 @@ console.log( v );
+
+#### stdlib_strided_dnanvarianceyc( N, correction, \*X, strideX )
+
+Computes the [variance][variance] of a double-precision floating-point strided array `x` ignoring `NaN` values and using a one-pass algorithm proposed by Youngs and Cramer.
```
```suggestion
Computes the [variance][variance] of a double-precision floating-point strided array, ignoring `NaN` values and using a one-pass algorithm proposed by Youngs and Cramer.
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dnanvarianceyc/README.md:
```diff
@@ -222,6 +220,125 @@ console.log( v );
+
+#### stdlib_strided_dnanvarianceyc_ndarray( N, correction, \*X, strideX, offsetX )
+
+Computes the [variance][variance] of a double-precision floating-point strided array ignoring `NaN` values and using a one-pass algorithm proposed by Youngs and Cramer and alternative indexing semantics.
```
```suggestion
Computes the [variance][variance] of a double-precision floating-point strided array, ignoring `NaN` values and using a one-pass algorithm proposed by Youngs and Cramer and alternative indexing semantics.
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dnanvarianceyc/README.md:
```diff
@@ -98,7 +98,7 @@ The use of the term `n-1` is commonly referred to as Bessel's correction. Note,
-#### dnanvarianceyc( N, correction, x, stride )
+#### dnanvarianceyc( N, correction, x, strideX )
 
 Computes the [variance][variance] of a double-precision floating-point strided array `x` ignoring `NaN` values and using a one-pass algorithm proposed by Youngs and Cramer.
```
```suggestion
Computes the [variance][variance] of a double-precision floating-point strided array, ignoring `NaN` values and using a one-pass algorithm proposed by Youngs and Cramer.
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dnanvarianceyc/README.md:
```diff
@@ -116,39 +116,36 @@ The function has the following parameters:
-#### dnanvarianceyc.ndarray( N, correction, x, stride, offset )
+#### dnanvarianceyc.ndarray( N, correction, x, strideX, offsetX )
 
 Computes the [variance][variance] of a double-precision floating-point strided array ignoring `NaN` values and using a one-pass algorithm proposed by Youngs and Cramer and alternative indexing semantics.
```
```suggestion
Computes the [variance][variance] of a double-precision floating-point strided array, ignoring `NaN` values and using a one-pass algorithm proposed by Youngs and Cramer and alternative indexing semantics.
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
